File: sundial.cpp
// sundial.cpp
#include <iostream>
#include <string>
#include <vector>
#include <cmath>
#include <ctime>
#include <iomanip>
#include <fstream>
#include <sstream>
#include <json/json.h> // using jsoncpp

using namespace std;

const double DEG = M_PI / 180.0;
const double RAD = 180.0 / M_PI;

struct Result {
    double altitude, azimuth, shadow_az, solar_time, dec, hour_angle;
    bool has_shadow_len;
    double shadow_len;
    bool above_horizon;
};

class Sundial {
private:
    double lat, lon, height, tz_offset;
    time_t date_time;
    time_t utc_time;

    double julianDay(time_t t) {
        tm* tm = gmtime(&t);
        int y = tm->tm_year + 1900;
        int m = tm->tm_mon + 1;
        double d = tm->tm_mday + tm->tm_hour/24.0 + tm->tm_min/1440.0 + tm->tm_sec/86400.0;
        if (m <= 2) { y--; m += 12; }
        int A = y / 100;
        int B = 2 - A + A / 4;
        return floor(365.25 * (y + 4716)) + floor(30.6001 * (m + 1)) + d + B - 1524.5;
    }

    double solarDeclination(time_t t) {
        tm* tm = gmtime(&t);
        time_t start = timegm(&tm);
        tm tm_start = {0}; tm_start.tm_year = tm->tm_year; tm_start.tm_mon = 0; tm_start.tm_mday = 1;
        double n = difftime(t, timegm(&tm_start)) / 86400 + 1;
        return -23.44 * cos(DEG * (360.0/365.0) * (n + 10));
    }

    double equationOfTime(time_t t) {
        tm* tm = gmtime(&t);
        tm tm_start = {0}; tm_start.tm_year = tm->tm_year; tm_start.tm_mon = 0; tm_start.tm_mday = 1;
        double n = difftime(t, timegm(&tm_start)) / 86400 + 1;
        double B = (360.0 / 365.0) * (n - 81);
        double B_rad = B * DEG;
        return 9.87 * sin(2 * B_rad) - 7.53 * cos(B_rad) - 1.5 * sin(B_rad);
    }

public:
    Sundial(double lat, double lon, double height, double tz, const string& dateStr, const string& timeStr) {
        this->lat = lat * DEG;
        this->lon = lon;
        this->height = height;
        if (tz != 0) this->tz_offset = tz;
        else { this->tz_offset = -timezone / 3600; }
        time_t t = time(nullptr);
        if (!dateStr.empty()) {
            tm tm = {0}; strptime(dateStr.c_str(), "%Y-%m-%d", &tm);
            t = timegm(&tm);
        }
        if (!timeStr.empty()) {
            int h, m, s = 0;
            sscanf(timeStr.c_str(), "%d:%d:%d", &h, &m, &s);
            tm* tm = gmtime(&t);
            tm->tm_hour = h; tm->tm_min = m; tm->tm_sec = s;
            t = timegm(tm);
        }
        date_time = t;
        utc_time = t - (time_t)(tz_offset * 3600);
    }

    Result calculate() {
        double eot = equationOfTime(utc_time);
        tm* tm = gmtime(&utc_time);
        double utcHours = tm->tm_hour + tm->tm_min/60.0 + tm->tm_sec/3600.0;
        double solarTime = utcHours + lon/15.0 + eot/60.0;
        double hourAngle = (solarTime - 12) * 15 * DEG;
        double dec = solarDeclination(utc_time) * DEG;
        double sinAlt = sin(lat) * sin(dec) + cos(lat) * cos(dec) * cos(hourAngle);
        double altitude = asin(sinAlt);
        double azimuth = 0;
        if (altitude > 0.01) {
            double cosAz = (sin(dec) - sin(lat) * sinAlt) / (cos(lat) * cos(altitude) + 1e-9);
            azimuth = acos(max(-1.0, min(1.0, cosAz)));
            if (hourAngle > 0) azimuth = -azimuth + 2*M_PI;
            azimuth = fmod(azimuth + 2*M_PI, 2*M_PI);
        }
        Result res;
        res.altitude = altitude * RAD;
        res.azimuth = azimuth * RAD;
        res.has_shadow_len = altitude > 0.01;
        res.shadow_len = res.has_shadow_len ? height / tan(altitude) : 0;
        res.shadow_az = res.has_shadow_len ? fmod(azimuth + M_PI, 2*M_PI) * RAD : 0;
        res.solar_time = solarTime;
        res.dec = dec * RAD;
        res.hour_angle = hourAngle * RAD;
        res.above_horizon = altitude > 0;
        return res;
    }

    void printResult(const Result& res, bool shadow, bool color) {
        string c = color ? "\033[36m" : "", cY = color ? "\033[33m" : "", cG = color ? "\033[32m" : "", cB = color ? "\033[34m" : "", cR = color ? "\033[31m" : "", reset = color ? "\033[0m" : "";
        cout << c << "☀️ Расчёт солнечной тени:" << reset << endl;
        char buf[20]; strftime(buf, sizeof(buf), "%Y-%m-%d %H:%M:%S", gmtime(&date_time));
        cout << "  " << cY << "Дата/время:" << reset << " " << buf << endl;
        cout << "  " << cY << "Широта:" << reset << " " << fixed << setprecision(4) << lat/RAD << "°, Долгота: " << lon << "°" << endl;
        cout << "  " << cY << "Высота гномона:" << reset << " " << fixed << setprecision(2) << height << " м" << endl;
        if (res.above_horizon) {
            cout << "  " << cG << "Высота Солнца:" << reset << " " << fixed << setprecision(2) << res.altitude << "°" << endl;
            cout << "  " << cG << "Азимут Солнца:" << reset << " " << fixed << setprecision(2) << res.azimuth << "°" << endl;
            if (res.has_shadow_len) {
                cout << "  " << cB << "Длина тени:" << reset << " " << fixed << setprecision(2) << res.shadow_len << " м" << endl;
            } else {
                cout << "  " << c << "Тень бесконечна (Солнце в зените)" << reset << endl;
            }
            cout << "  " << cB << "Направление тени:" << reset << " " << fixed << setprecision(2) << res.shadow_az << "°" << endl;
            if (shadow) drawShadow(res);
        } else {
            cout << cR << "  Солнце под горизонтом (ночь)" << reset << endl;
        }
    }

    void drawShadow(const Result& res) {
        int size = 15, center = size/2;
        double scale = 2.0;
        double shadowLen = res.has_shadow_len ? min(res.shadow_len, 10.0) / scale : 0;
        double shadowAz = res.shadow_az * DEG;
        int x = (int)round(center + shadowLen * sin(shadowAz));
        int y = (int)round(center - shadowLen * cos(shadowAz));
        vector<vector<char>> grid(size, vector<char>(size, '░'));
        grid[center][center] = '╋';
        if (x >= 0 && x < size && y >= 0 && y < size && res.has_shadow_len) {
            int steps = max(abs(x-center), abs(y-center));
            for (int i = 0; i <= steps; i++) {
                int cx = (int)round(center + (double)(x-center) * i / steps);
                int cy = (int)round(center + (double)(y-center) * i / steps);
                if (cx >= 0 && cx < size && cy >= 0 && cy < size && grid[cy][cx] == '░')
                    grid[cy][cx] = '█';
            }
        }
        cout << "\n\033[36mВид сверху (N ↑):\033[0m" << endl;
        cout << "  ";
        for (int i = 0; i < size; i++) cout << setw(2) << i;
        cout << endl;
        for (int i = 0; i < size; i++) {
            cout << setw(2) << i << " ";
            for (int j = 0; j < size; j++) {
                char c = grid[i][j];
                if (c == '█') cout << "\033[32m█\033[0m";
                else if (c == '╋') cout << "\033[37m╋\033[0m";
                else cout << "\033[34m░\033[0m";
            }
            cout << endl;
        }
    }

    void exportJSON(const Result& res, const string& filename) {
        Json::Value root;
        root["time"] = (double)date_time;
        root["lat"] = lat/RAD;
        root["lon"] = lon;
        root["height"] = height;
        root["result"]["altitude"] = res.altitude;
        root["result"]["azimuth"] = res.azimuth;
        root["result"]["shadow_len"] = res.has_shadow_len ? res.shadow_len : 0;
        root["result"]["shadow_az"] = res.shadow_az;
        root["result"]["above_horizon"] = res.above_horizon;
        ofstream ofs(filename);
        ofs << root.toStyledString();
    }

    void exportCSV(const Result& res, const string& filename) {
        ofstream ofs(filename);
        ofs << "time,altitude,azimuth,shadow_len,shadow_az\n";
        char buf[20]; strftime(buf, sizeof(buf), "%Y-%m-%d %H:%M:%S", gmtime(&date_time));
        ofs << buf << "," << res.altitude << "," << res.azimuth << ","
            << (res.has_shadow_len ? to_string(res.shadow_len) : "") << "," << res.shadow_az << "\n";
    }

    vector<map<string, double>> daySeries() {
        vector<map<string, double>> results;
        time_t t = date_time;
        tm* tm = gmtime(&t);
        tm->tm_hour = 0; tm->tm_min = 0; tm->tm_sec = 0;
        t = timegm(tm);
        for (int h = 0; h < 24; h++) {
            time_t dt = t + h * 3600;
            Sundial tmp(lat/RAD, lon, height, tz_offset, "", "");
            tmp.date_time = dt;
            tmp.utc_time = dt - (time_t)(tz_offset * 3600);
            Result res = tmp.calculate();
            map<string, double> entry;
            entry["time"] = dt;
            entry["altitude"] = res.altitude;
            entry["azimuth"] = res.azimuth;
            entry["shadow_len"] = res.has_shadow_len ? res.shadow_len : -1;
            entry["shadow_az"] = res.shadow_az;
            entry["above_horizon"] = res.above_horizon ? 1 : 0;
            results.push_back(entry);
        }
        return results;
    }
};

int main(int argc, char* argv[]) {
    double lat = 0, lon = 0, height = 1.0, tz = 0;
    string dateStr, timeStr, jsonFile, csvFile;
    bool shadow = false, day = false, color = false;

    for (int i = 1; i < argc; ++i) {
        string arg = argv[i];
        if (arg == "--lat" && i+1 < argc) lat = stod(argv[++i]);
        else if (arg == "--lon" && i+1 < argc) lon = stod(argv[++i]);
        else if (arg == "--date" && i+1 < argc) dateStr = argv[++i];
        else if (arg == "--time" && i+1 < argc) timeStr = argv[++i];
        else if (arg == "--height" && i+1 < argc) height = stod(argv[++i]);
        else if (arg == "--tz" && i+1 < argc) tz = stod(argv[++i]);
        else if (arg == "--shadow") shadow = true;
        else if (arg == "--day") day = true;
        else if (arg == "--json" && i+1 < argc) jsonFile = argv[++i];
        else if (arg == "--csv" && i+1 < argc) csvFile = argv[++i];
        else if (arg == "--color") color = true;
    }

    if (lat == 0 && lon == 0) {
        cerr << "Ошибка: --lat и --lon обязательны" << endl;
        return 1;
    }
    color = color || isatty(fileno(stdout));

    Sundial sundial(lat, lon, height, tz, dateStr, timeStr);

    if (day) {
        auto results = sundial.daySeries();
        char buf[20]; strftime(buf, sizeof(buf), "%Y-%m-%d", gmtime(&sundial.date_time));
        cout << "\033[36m📊 Тень на " << buf << " (почасово):\033[0m" << endl;
        for (auto& r : results) {
            if (r["above_horizon"] > 0.5) {
                string sl = r["shadow_len"] >= 0 ? to_string((int)r["shadow_len"]) : "∞";
                char tb[20]; time_t t = (time_t)r["time"]; strftime(tb, sizeof(tb), "%H:%M", gmtime(&t));
                cout << "  " << tb << "  Alt: " << setw(5) << fixed << setprecision(1) << r["altitude"]
                     << "°  Тень: " << setw(6) << sl << " м  Направление: " << setw(6) << fixed << setprecision(1) << r["shadow_az"] << "°" << endl;
            } else {
                char tb[20]; time_t t = (time_t)r["time"]; strftime(tb, sizeof(tb), "%H:%M", gmtime(&t));
                cout << "  " << tb << "  Ночь" << endl;
            }
        }
        if (!jsonFile.empty()) {
            Json::Value root(Json::arrayValue);
            for (auto& r : results) {
                Json::Value item;
                item["time"] = (double)r["time"];
                item["altitude"] = r["altitude"];
                item["azimuth"] = r["azimuth"];
                item["shadow_len"] = r["shadow_len"];
                item["shadow_az"] = r["shadow_az"];
                item["above_horizon"] = r["above_horizon"] > 0.5;
                root.append(item);
            }
            ofstream ofs(jsonFile);
            ofs << root.toStyledString();
            cout << "Экспортировано в " << jsonFile << endl;
        }
        if (!csvFile.empty()) {
            ofstream ofs(csvFile);
            ofs << "time,altitude,azimuth,shadow_len,shadow_az\n";
            for (auto& r : results) {
                char tb[20]; time_t t = (time_t)r["time"]; strftime(tb, sizeof(tb), "%Y-%m-%d %H:%M:%S", gmtime(&t));
                ofs << tb << "," << r["altitude"] << "," << r["azimuth"] << ","
                    << (r["shadow_len"] >= 0 ? to_string(r["shadow_len"]) : "") << "," << r["shadow_az"] << "\n";
            }
            cout << "Экспортировано в " << csvFile << endl;
        }
    } else {
        Result res = sundial.calculate();
        sundial.printResult(res, shadow, color);
        if (!jsonFile.empty()) {
            sundial.exportJSON(res, jsonFile);
            cout << "\033[32mЭкспортировано в " << jsonFile << "\033[0m" << endl;
        }
        if (!csvFile.empty()) {
            sundial.exportCSV(res, csvFile);
            cout << "\033[32mЭкспортировано в " << csvFile << "\033[0m" << endl;
        }
    }
    return 0;
}

File: Sundial.cs
// Sundial.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Sundial
{
    class Program
    {
        static void Main(string[] args)
        {
            var opts = ParseArgs(args);
            var sundial = new Sundial(opts);
            if (opts.Day)
            {
                var results = sundial.DaySeries();
                Console.WriteLine($"\u001B[36m📊 Тень на {sundial.DateTime:yyyy-MM-dd} (почасово):\u001B[0m");
                foreach (var r in results)
                {
                    if ((bool)r["above_horizon"])
                    {
                        string shadowLen = r["shadow_len"] != null ? $"{r["shadow_len"]:F2}" : "∞";
                        Console.WriteLine($"  {r["time"]:HH:mm}  Alt: {r["altitude"]:F1}°  Тень: {shadowLen,6} м  Направление: {r["shadow_az"]:F1}°");
                    }
                    else
                    {
                        Console.WriteLine($"  {r["time"]:HH:mm}  Ночь");
                    }
                }
                if (opts.Json != null)
                {
                    string json = JsonSerializer.Serialize(results, new JsonSerializerOptions { WriteIndented = true });
                    File.WriteAllText(opts.Json, json);
                    Console.WriteLine($"Экспортировано в {opts.Json}");
                }
                if (opts.Csv != null)
                {
                    using var sw = new StreamWriter(opts.Csv);
                    sw.WriteLine("time,altitude,azimuth,shadow_len,shadow_az");
                    foreach (var r in results)
                    {
                        sw.WriteLine($"{r["time"]},{r["altitude"]},{r["azimuth"]},{r["shadow_len"]},{r["shadow_az"]}");
                    }
                    Console.WriteLine($"Экспортировано в {opts.Csv}");
                }
            }
            else
            {
                var res = sundial.Calculate();
                sundial.PrintResult(res);
                if (opts.Json != null)
                {
                    sundial.ExportJson(res, opts.Json);
                    Console.WriteLine($"\u001B[32mЭкспортировано в {opts.Json}\u001B[0m");
                }
                if (opts.Csv != null)
                {
                    sundial.ExportCsv(res, opts.Csv);
                    Console.WriteLine($"\u001B[32mЭкспортировано в {opts.Csv}\u001B[0m");
                }
            }
        }

        static Options ParseArgs(string[] args)
        {
            var opts = new Options();
            for (int i = 0; i < args.Length; i++)
            {
                switch (args[i])
                {
                    case "--lat": opts.Lat = double.Parse(args[++i], CultureInfo.InvariantCulture); break;
                    case "--lon": opts.Lon = double.Parse(args[++i], CultureInfo.InvariantCulture); break;
                    case "--date": opts.Date = args[++i]; break;
                    case "--time": opts.Time = args[++i]; break;
                    case "--height": opts.Height = double.Parse(args[++i], CultureInfo.InvariantCulture); break;
                    case "--tz": opts.Tz = double.Parse(args[++i], CultureInfo.InvariantCulture); break;
                    case "--shadow": opts.Shadow = true; break;
                    case "--day": opts.Day = true; break;
                    case "--json": opts.Json = args[++i]; break;
                    case "--csv": opts.Csv = args[++i]; break;
                }
            }
            return opts;
        }

        class Options
        {
            public double Lat { get; set; }
            public double Lon { get; set; }
            public string Date { get; set; }
            public string Time { get; set; }
            public double Height { get; set; } = 1.0;
            public double? Tz { get; set; }
            public bool Shadow { get; set; }
            public bool Day { get; set; }
            public string Json { get; set; }
            public string Csv { get; set; }
        }

        class Result
        {
            public double Altitude { get; set; }
            public double Azimuth { get; set; }
            public double? ShadowLen { get; set; }
            public double ShadowAz { get; set; }
            public double SolarTime { get; set; }
            public double Dec { get; set; }
            public double HourAngle { get; set; }
            public bool AboveHorizon { get; set; }
        }

        class Sundial
        {
            private const double DEG = Math.PI / 180;
            private const double RAD = 180 / Math.PI;

            private readonly double lat, lon, height, tzOffset;
            public DateTime DateTime { get; private set; }
            private DateTime utcTime;

            public Sundial(Options opts)
            {
                lat = opts.Lat * DEG;
                lon = opts.Lon;
                height = opts.Height;
                tzOffset = opts.Tz ?? -(double)TimeZoneInfo.Local.GetUtcOffset(DateTime.UtcNow).TotalHours;
                DateTime dt = DateTime.UtcNow;
                if (opts.Date != null)
                    dt = DateTime.ParseExact(opts.Date, "yyyy-MM-dd", CultureInfo.InvariantCulture);
                if (opts.Time != null)
                {
                    var parts = opts.Time.Split(':');
                    dt = dt.Date.AddHours(int.Parse(parts[0])).AddMinutes(int.Parse(parts[1]));
                    if (parts.Length > 2) dt = dt.AddSeconds(int.Parse(parts[2]));
                }
                DateTime = dt;
                utcTime = dt.AddHours(-tzOffset);
            }

            private double JulianDay(DateTime dt)
            {
                int y = dt.Year;
                int m = dt.Month;
                double d = dt.Day + dt.Hour/24.0 + dt.Minute/1440.0 + dt.Second/86400.0;
                if (m <= 2) { y--; m += 12; }
                int A = y / 100;
                int B = 2 - A + A / 4;
                return Math.Floor(365.25 * (y + 4716)) + Math.Floor(30.6001 * (m + 1)) + d + B - 1524.5;
            }

            private double SolarDeclination(DateTime dt)
            {
                var start = new DateTime(dt.Year, 1, 1, 0, 0, 0, DateTimeKind.Utc);
                double n = (dt - start).TotalDays + 1;
                return -23.44 * Math.Cos(DEG * (360.0/365.0) * (n + 10));
            }

            private double EquationOfTime(DateTime dt)
            {
                var start = new DateTime(dt.Year, 1, 1, 0, 0, 0, DateTimeKind.Utc);
                double n = (dt - start).TotalDays + 1;
                double B = (360.0 / 365.0) * (n - 81);
                double B_rad = B * DEG;
                return 9.87 * Math.Sin(2 * B_rad) - 7.53 * Math.Cos(B_rad) - 1.5 * Math.Sin(B_rad);
            }

            public Result Calculate()
            {
                double eot = EquationOfTime(utcTime);
                double utcHours = utcTime.Hour + utcTime.Minute/60.0 + utcTime.Second/3600.0;
                double solarTime = utcHours + lon/15.0 + eot/60.0;
                double hourAngle = (solarTime - 12) * 15 * DEG;
                double dec = SolarDeclination(utcTime) * DEG;
                double latRad = lat;
                double sinAlt = Math.Sin(latRad) * Math.Sin(dec) + Math.Cos(latRad) * Math.Cos(dec) * Math.Cos(hourAngle);
                double altitude = Math.Asin(sinAlt);
                double azimuth = 0;
                if (altitude > 0.01)
                {
                    double cosAz = (Math.Sin(dec) - Math.Sin(latRad) * sinAlt) / (Math.Cos(latRad) * Math.Cos(altitude) + 1e-9);
                    azimuth = Math.Acos(Math.Max(-1, Math.Min(1, cosAz)));
                    if (hourAngle > 0) azimuth = -azimuth + 2*Math.PI;
                    azimuth = (azimuth + 2*Math.PI) % (2*Math.PI);
                }
                double? shadowLen = altitude > 0.01 ? height / Math.Tan(altitude) : (double?)null;
                double shadowAz = altitude > 0.01 ? (azimuth + Math.PI) % (2*Math.PI) * RAD : 0;
                return new Result
                {
                    Altitude = altitude * RAD,
                    Azimuth = azimuth * RAD,
                    ShadowLen = shadowLen,
                    ShadowAz = shadowAz,
                    SolarTime = solarTime,
                    Dec = dec * RAD,
                    HourAngle = hourAngle * RAD,
                    AboveHorizon = altitude > 0
                };
            }

            public void PrintResult(Result res)
            {
                bool color = !Console.IsOutputRedirected;
                string c = color ? "\u001B[36m" : "", cY = color ? "\u001B[33m" : "", cG = color ? "\u001B[32m" : "", cB = color ? "\u001B[34m" : "", cR = color ? "\u001B[31m" : "", reset = color ? "\u001B[0m" : "";
                Console.WriteLine($"{c}☀️ Расчёт солнечной тени:{reset}");
                Console.WriteLine($"  {cY}Дата/время:{reset} {DateTime:yyyy-MM-dd HH:mm:ss}");
                Console.WriteLine($"  {cY}Широта:{reset} {lat/RAD:F4}°, Долгота: {lon:F4}°");
                Console.WriteLine($"  {cY}Высота гномона:{reset} {height:F2} м");
                if (res.AboveHorizon)
                {
                    Console.WriteLine($"  {cG}Высота Солнца:{reset} {res.Altitude:F2}°");
                    Console.WriteLine($"  {cG}Азимут Солнца:{reset} {res.Azimuth:F2}°");
                    if (res.ShadowLen.HasValue)
                        Console.WriteLine($"  {cB}Длина тени:{reset} {res.ShadowLen.Value:F2} м");
                    else
                        Console.WriteLine($"  {c}Тень бесконечна (Солнце в зените){reset}");
                    Console.WriteLine($"  {cB}Направление тени:{reset} {res.ShadowAz:F2}°");
                    if (opts.Shadow) DrawShadow(res);
                }
                else
                {
                    Console.WriteLine($"{cR}  Солнце под горизонтом (ночь){reset}");
                }
            }

            private void DrawShadow(Result res)
            {
                int size = 15, center = size / 2;
                double scale = 2.0;
                double shadowLen = res.ShadowLen.HasValue ? Math.Min(res.ShadowLen.Value, 10) / scale : 0;
                double shadowAz = res.ShadowAz * DEG;
                int x = (int)Math.Round(center + shadowLen * Math.Sin(shadowAz));
                int y = (int)Math.Round(center - shadowLen * Math.Cos(shadowAz));
                char[][] grid = new char[size][];
                for (int i = 0; i < size; i++) { grid[i] = new char[size]; for (int j = 0; j < size; j++) grid[i][j] = '░'; }
                grid[center][center] = '╋';
                if (x >= 0 && x < size && y >= 0 && y < size && res.ShadowLen.HasValue)
                {
                    int steps = Math.Max(Math.Abs(x-center), Math.Abs(y-center));
                    for (int i = 0; i <= steps; i++)
                    {
                        int cx = (int)Math.Round(center + (double)(x-center) * i / steps);
                        int cy = (int)Math.Round(center + (double)(y-center) * i / steps);
                        if (cx >= 0 && cx < size && cy >= 0 && cy < size && grid[cy][cx] == '░')
                            grid[cy][cx] = '█';
                    }
                }
                Console.WriteLine("\n\u001B[36mВид сверху (N ↑):\u001B[0m");
                Console.Write("  ");
                for (int i = 0; i < size; i++) Console.Write($"{i,2}");
                Console.WriteLine();
                for (int i = 0; i < size; i++)
                {
                    Console.Write($"{i,2} ");
                    for (int j = 0; j < size; j++)
                    {
                        char c = grid[i][j];
                        Console.Write(c == '█' ? "\u001B[32m█\u001B[0m" : c == '╋' ? "\u001B[37m╋\u001B[0m" : "\u001B[34m░\u001B[0m");
                    }
                    Console.WriteLine();
                }
            }

            public void ExportJson(Result res, string filename)
            {
                var data = new { time = DateTime, lat = lat/RAD, lon, height, result = res };
                string json = JsonSerializer.Serialize(data, new JsonSerializerOptions { WriteIndented = true });
                File.WriteAllText(filename, json);
            }

            public void ExportCsv(Result res, string filename)
            {
                using var sw = new StreamWriter(filename);
                sw.WriteLine("time,altitude,azimuth,shadow_len,shadow_az");
                sw.WriteLine($"{DateTime:o},{res.Altitude:F2},{res.Azimuth:F2},{res.ShadowLen},{res.ShadowAz:F2}");
            }

            public List<Dictionary<string, object>> DaySeries()
            {
                var results = new List<Dictionary<string, object>>();
                var dt = DateTime.Date;
                while (dt.Hour < 24)
                {
                    var tmp = new Sundial(new Options { Lat = lat/RAD, Lon = lon, Height = height, Tz = tzOffset });
                    tmp.DateTime = dt;
                    tmp.utcTime = dt.AddHours(-tzOffset);
                    var res = tmp.Calculate();
                    results.Add(new Dictionary<string, object>
                    {
                        ["time"] = dt,
                        ["altitude"] = res.Altitude,
                        ["azimuth"] = res.Azimuth,
                        ["shadow_len"] = res.ShadowLen,
                        ["shadow_az"] = res.ShadowAz,
                        ["above_horizon"] = res.AboveHorizon
                    });
                    dt = dt.AddHours(1);
                }
                return results;
            }
        }
    }
}
